test/e2e/storage/non_graceful_node_shutdown_scale.go

```diff
@@ -0,0 +1,257 @@
+			taintNew := v1.Taint{
+				Key:    "STSPods",
+				Effect: v1.TaintEffectNoSchedule,
+			}
+			for _, nodeName := range nodesToBeTainted {
+				e2enode.AddOrUpdateTaintOnNode(ctx, c, nodeName.Name, taintNew)
+			}
+
+			// Create STS with 'N/2' replicas ( 'N' = Number of nodes in the k8s cluster ).
+			replicaCount := int32(MinNodeRequired / 2)
+			stsName := "sts-app-gcepd"
+			podLabels := map[string]string{"app": stsName}
+			podsList := createAndVerifySTS(ctx, scName, stsName, ns, &replicaCount, podLabels, c)
+			oldNodeNameList := make([]string, 0)
+			for _, pod := range podsList {
+				oldNodeNameList = append(oldNodeNameList, pod.Spec.NodeName)
+			}
+
+			// stop the kubelet on nodes where the sts pods were scheduled.
+			for _, pod := range podsList {
+				ginkgo.By("Stopping the kubelet non gracefully for pod" + pod.Name)
+				utils.KubeletCommand(ctx, utils.KStop, c, &pod)
+				ginkgo.DeferCleanup(utils.KubeletCommand, utils.KStart, c, pod)
+			}
+
+			// tainting the shutdown nodes with `out-of-service` taint
+			taint := v1.Taint{
+				Key:    v1.TaintNodeOutOfService,
+				Effect: v1.TaintEffectNoExecute,
+			}
+			for _, oldNodeName := range oldNodeNameList {
+				ginkgo.By("Adding out of service taint on node " + oldNodeName)
+				// taint this node as out-of-service node
+				e2enode.AddOrUpdateTaintOnNode(ctx, c, oldNodeName, taint)
+				ginkgo.DeferCleanup(e2enode.RemoveTaintOffNode, c, oldNodeName, taint)
+			}
+
+			ginkgo.By(fmt.Sprintf("Checking if the pods of sts %s got rescheduled to a new node", stsName))
+			// building field selector such that pods on a particular list of nodes are not selected.
+			nodeFieldSelector := make([]fields.Selector, 0)
+			for _, v := range oldNodeNameList {
+				nodeFieldSelector = append(nodeFieldSelector, fields.OneTermNotEqualSelector("spec.nodeName", v))
+			}
+			labelSelectorStr := labels.SelectorFromSet(podLabels).String()
+			podListOpts := metav1.ListOptions{
+				LabelSelector: labelSelectorStr,
+				FieldSelector: fields.AndSelectors(nodeFieldSelector...).String(),
+			}
+			_, err = e2epod.WaitForPods(ctx, c, ns, podListOpts, e2epod.Range{MinMatching: int(replicaCount)}, framework.PodStartTimeout, "be running and ready", e2epod.RunningReady)
+			framework.ExpectNoError(err)
+
+			// Bring the nodes back online.
+			for _, pod := range podsList {
+				utils.KubeletCommand(ctx, utils.KStart, c, &pod)
+			}
+
+			// Remove the `out-of-service` taint on the nodes.
+			for _, oldNodeName := range oldNodeNameList {
+				e2enode.RemoveTaintOffNode(ctx, c, oldNodeName, taint)
+			}
+
+			// Verify that pods gets scheduled to the older nodes that was terminated non gracefully and now
+			// is back online
+			newSTSName := "sts-app-gcepd-new"
+			newPodLabels := map[string]string{"app": newSTSName}
+			newReplicaCount := int32(MinNodeRequired)
+			createAndVerifySTS(ctx, scName, newSTSName, ns, &newReplicaCount, newPodLabels, c)
+		})
+	})
+})
```

Comment:
Do these pods and PVCs get cleaned up at the end of testing?

Reply:
As per the current implementation, it does not. I will add that logic. Also, is the cleanup required?

Reply:
Yes, typically the resources created for e2e tests should be cleaned up at the end of the tests.

Reply:
ACK. Making the changes.